source/distributed/tria_base.cc

```diff
@@ -0,0 +1,184 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2008 - 2013 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+#include <deal.II/base/utilities.h>
+#include <deal.II/base/memory_consumption.h>
+#include <deal.II/base/logstream.h>
+#include <deal.II/lac/sparsity_tools.h>
+#include <deal.II/lac/sparsity_pattern.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/grid/tria_iterator.h>
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/distributed/tria_base.h>
+
+
+#include <algorithm>
+#include <numeric>
+#include <iostream>
+#include <fstream>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace parallel
+{
+
+#ifdef DEAL_II_WITH_MPI
+  template <int dim, int spacedim>
+  Triangulation<dim,spacedim>::Triangulation (MPI_Comm mpi_communicator,
+                                              const typename dealii::Triangulation<dim,spacedim>::MeshSmoothing smooth_grid,
+                                              const bool check_for_distorted_cells)
+    :
+    dealii::Triangulation<dim,spacedim>(smooth_grid,check_for_distorted_cells),
+    mpi_communicator (Utilities::MPI::
+                      duplicate_communicator(mpi_communicator)),
+    my_subdomain (Utilities::MPI::this_mpi_process (this->mpi_communicator)),
+    n_subdomains(Utilities::MPI::n_mpi_processes(mpi_communicator))
+  {
+    number_cache.n_locally_owned_active_cells.resize (n_subdomains);
+  };
+#else
+  template <int dim, int spacedim>
+  Triangulation<dim,spacedim>::Triangulation()
+  {
+    Assert (false, ExcNotImplemented());
+  }
+
+#endif
+
+  template <int dim, int spacedim>
+  Triangulation<dim,spacedim>::~Triangulation ()
+  {};
```

Comment:
here is the place where you need to delete the duplicated MPI communicator (given that you duplicated it above in the constructor).